ffi/src/ATen/GC.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE EmptyDataDecls #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module ATen.GC where
+
+import Control.Exception.Safe (catch,throwIO)
+import Data.List (isPrefixOf)
+import Language.C.Inline.Cpp.Exceptions (CppException(..))
+import System.Mem (performGC)
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.Async
+import System.SysInfo
+
+retryWithGC' :: Int -> IO a -> IO a
+retryWithGC' count func =
+  func `catch` \a@(CppStdException message) ->
+    if isPrefixOf msgOutOfMemory message
+    then
+      if count <= 0
+      then throwIO $ CppStdException $ "Too many calls to performGC, " ++ message
+      else do
+        performGC
```

Comment:
Are you sure it is safe to call `performGC` from inside the `unsafePerformIO` block? We will be calling those functions in unsafe ways

Reply:
I do not know safe or not.  I ask the question for GHC.
https://gitlab.haskell.org/ghc/ghc/issues/16871